Assets/Mirror/Runtime/NetworkReader.cs

```diff
@@ -10,51 +12,172 @@ namespace Mirror
+        public short ReadInt16() => (short)ReadUInt16();
+        public ushort ReadUInt16()
+        {
+            ushort value = 0;
+            value |= ReadByte();
+            value |= (ushort)(ReadByte() << 8);
+            return value;
+        }
+        public int ReadInt32() => (int)ReadUInt32();
+        public uint ReadUInt32()
+        {
+            uint value = 0;
+            value |= ReadByte();
+            value |= (uint)(ReadByte() << 8);
+            value |= (uint)(ReadByte() << 16);
+            value |= (uint)(ReadByte() << 24);
+            return value;
+        }
+        public long ReadInt64() => (long)ReadUInt64();
+        public ulong ReadUInt64()
+        {
+            ulong value = 0;
+            ulong other = ReadByte();
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 8;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 16;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 24;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 32;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 40;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 48;
+            value |= other;
+
+            other = ((ulong)ReadByte()) << 56;
+            value |= other;
+            return value;
+        }
+        public decimal ReadDecimal()
+        {
+            int[] bits = new int[4];
+
+            bits[0] = ReadInt32();
+            bits[1] = ReadInt32();
+            bits[2] = ReadInt32();
+            bits[3] = ReadInt32();
+
+            return new decimal(bits);
+        }
+        public float ReadSingle()
+        {
+            uint value = ReadUInt32();
+            return FloatConversion.ToSingle(value);
+        }
+        public double ReadDouble()
+        {
+            ulong value = ReadUInt64();
+            return FloatConversion.ToDouble(value);
+        }
 
         // note: this will throw an ArgumentException if an invalid utf8 string is sent
         // null support, see NetworkWriter
-        public string ReadString() => ReadBoolean() ? reader.ReadString() : null;
+        public string ReadString()
+        {
+            // isNull?
+            if (ReadBoolean())
+            {
+                // read number of bytes
+                UInt16 numBytes = ReadUInt16();
+                if (numBytes == 0)
+                    return "";
+
+                // make sure it's within limits to avoid allocation attacks etc.
+                if (numBytes >= stringBuffer.Length)
+                {
+                    throw new IndexOutOfRangeException("ReadString() too long: " + numBytes + ". Limit is: " + stringBuffer.Length);
+                }
+
+                // read the bytes
+                ReadBytes(stringBuffer, numBytes);
+
+                // convert to string via encoding
+                char[] chars = encoding.GetChars(stringBuffer, 0, numBytes);
```

Comment:
```suggestion
                return encoding.GetString(stringBuffer, 0, numBytes);
```

Avoid the double allocation